File: python/meetup/meetup.py
from datetime import date
from calendar import monthrange


def meetup_day(year, month, weekday, order):
    fstWeekDay = first_weekday(year, month, weekday)
    if order == "teenth":
        day = 13 + (fstWeekDay + 1 % 13)
        return date(year, month, day)

    if order == "last":
        day = last_weekday(year, month, weekday)
        return date(year, month, day)

    week = int(order[0])
    daysToMeetup = 7 * (week - 1)
    day = fstWeekDay + daysToMeetup
    return date(year, month, day)


def first_weekday(year, month, weekday):
    firstDayOfMonth = date(year, month, 1)
    diffWeekdays = weekday_to_int(weekday) - firstDayOfMonth.isoweekday()
    return diffWeekdays % 7 + 1


def last_weekday(year, month, weekday):
    fstWeekDay = first_weekday(year, month, weekday)
    if fstWeekDay + 28 in monthrange(year, month):
        return fstWeekDay + 28
    return fstWeekDay + 21


def weekday_to_int(weekday):
    try:
        return weekday_to_int_table[weekday]
    except:
        raise Exception(weekday + " is not a valid weekday")

weekday_to_int_table = {"Monday": 1, "Tuesday": 2, "Wednesday": 3,
                        "Thursday": 4, "Friday": 5, "Saturday": 6, "Sunday": 7}

File: haskell/meetup/Meetup.hs
module Meetup(Weekday(..), Schedule(..), meetupDay) where

import Data.Tuple.Extra(thd3)
import qualified Data.Time.Calendar as Calendar
import Data.Time.Calendar.WeekDate(toWeekDate)

type Year = Integer
type Month = Int
type Day = Int
data Weekday = Monday | Tuesday | Wednesday | Thursday |
               Friday | Saturday | Sunday deriving (Eq, Show, Read, Enum)

weekdayToIso :: Weekday -> Day
weekdayToIso = succ . fromEnum
isoToWeekday :: Day -> Weekday
isoToWeekday = toEnum . pred

data Schedule = First | Second | Third | Fourth | Teenth | Last
    deriving (Eq, Show, Read)

meetupDay :: Schedule -> Weekday -> Year -> Month -> Calendar.Day
meetupDay sch wd yy mm
    | when First  = toGregrorianDay fstWd
    | when Second = toGregrorianDay ( 7 + fstWd)
    | when Third  = toGregrorianDay (14 + fstWd)
    | when Fourth = toGregrorianDay (21 + fstWd)
    | when Teenth =
        if fstWd <= weekdayToIso Friday then
            meetupDay Third wd yy mm
        else
            meetupDay Second wd yy mm
    | otherwise = toGregrorianDay (lastWeekday wd yy mm)
    where
        when = (== sch)
        toGregrorianDay = Calendar.fromGregorian yy mm
        fstWd = firstWeekday wd yy mm

weekdayFirstDay :: Year -> Month -> Day
weekdayFirstDay yy mm = thd3 (toWeekDate (Calendar.fromGregorian yy mm 01))

firstWeekday :: Weekday -> Year -> Month -> Day
firstWeekday wd yy mm = ((weekdayToIso wd - weekdayFirstDay yy mm) `mod` 7) + 1

lastWeekday :: Weekday -> Year -> Month -> Day
lastWeekday wd yy mm =
    if fstWd + 28 > lastMonthDay then
        fstWd + 21
    else fstWd + 28
    where
        fstWd = firstWeekday wd yy mm
        lastMonthDay = Calendar.gregorianMonthLength yy mm
